lib/src/lifecycle_module.dart

```diff
@@ -264,18 +247,100 @@ abstract class LifecycleModule extends SimpleModule
   /// Whether the module is currently unloading.
   bool get isUnloading => _state == LifecycleState.unloading;
 
-  @override
-  StreamSubscription<T> listenToStream<T>(
-          Stream<T> stream, void onData(T event),
-          {Function onError, void onDone(), bool cancelOnError}) =>
-      _disposableProxy.listenToStream(stream, onData,
-          onError: onError, onDone: onDone, cancelOnError: cancelOnError);
-
   //--------------------------------------------------------
   // Public methods that can be used directly to trigger
   // module lifecycle / check current lifecycle state
   //--------------------------------------------------------
 
+  /// Public method to dispose of this module.
+  ///
+  /// If the module has only been instantiated and has not yet started loading
+  /// or been loaded, then this will immediately dispose of the module.
+  ///
+  /// If the module has already started loading, has loaded, or is in any other
+  /// "loaded" state (suspending, suspended, resuming), then this will attempt
+  /// to unload the module before disposing.
+  ///
+  /// If the module has already started unloading, this will wait for that
+  /// transition before disposing.
+  ///
+  /// If the module has already started disposing or has disposed, then this
+  /// will return the [Future] from [didDispose]. (An unloaded module will have
+  /// already started or finished disposal).
+  ///
+  /// In any of these cases where an unload is attemped prior to disposal, a
+  /// failure during unload will be caught and logged, but will not stop
+  /// disposal. A module who cancels unload via [onShouldUnload] or who throws
+  /// during [onUnload] will still be disposed.
+  ///
+  /// In short, calling [dispose] forces the disposal of this module regardless
+  /// of its current state and regardless of its ability to unload successfully.
+  ///
+  /// If the modules unload is canceled or if an error is thrown during a
+  /// lifecycle handler like onUnload as a part of this disposal process, they
+  /// will still be available via their corresponding lifecycle event streams
+  /// (e.g. [didUnload]).
+  ///
+  /// The [Future] returned from this method will resolve when disposal has
+  /// completed and will only resolve with an error if one is thrown during
+  /// [onDispose].
+  @mustCallSuper
+  @override
+  Future<Null> dispose() async {
+    if (isDisposedOrDisposing) {
+      return didDispose;
+    }
+
+    if (_customDisposal != null) {
+      return _customDisposal.future;
+    }
+
+    _customDisposal = new Completer<Null>();
+
+    if (isInstantiated) {
+      // If the module has not yet been loaded, then we can go straight to
+      // disposal.
+      _customDisposal.complete(super.dispose().then((_) => didDispose));
+      return _customDisposal.future;
+    }
```

Comment:
Testing if it is dispose or disposing and returning a single future for multiple calls to `dispose` is handled by the `super.dispose` [implementation](https://github.com/Workiva/w_common/blob/master/lib/src/common/disposable.dart#L323-L336). Why duplicate it here?

I may be dense and am just missing a critical detail.

```diff
   Future<Null> dispose() async {
-    if (isDisposedOrDisposing) {
-      return didDispose;
-    }
-
-    if (_customDisposal != null) {
-      return _customDisposal.future;
-    }
-
-    _customDisposal = new Completer<Null>();
-
     if (isInstantiated) {
-      // If the module has not yet been loaded, then we can go straight to
-      // disposal.
-      _customDisposal.complete(super.dispose().then((_) => didDispose));
-      return _customDisposal.future;
+      return super.dispose();
     }
```

Reply:
I agree this would be much simpler, but unfortunately I have to handle some manual disposal logic before I can pass it off to `super.dispose()`. When `Module.dispose()` is called, we need to attempt to unload the module before moving on to disposing it. That unload process emits events which requires that `StreamController`s still be available. Due to the order in which disposal events happen in `Disposable.dispose()` (disposables are torn down and _then_ `onDispose()` is called), I have to override `dispose()` to do this instead of just overriding `onDispose()`. This also means that I need a separate piece to track this custom disposal phase since `isDisposedOrDisposing` won't be true until we actually call `super.dispose()` which doesn't happen until after attempting to unload the module.

I did quickly try moving the `onDispose()` step up in the `Disposable.dispose()` logic to see if I could then just override `onDispose()` and get rid of some of this custom logic, but I hit a bunch of stream-related errors so I would have to dig into that more to see if it's a workable solution.

Reply:
I see what you mean. I was hoping see duplicated Disposable logic removed 🤕